### ISSUE TITLE
Deserialization of the message body not happening correctly in RedisOperationsSessionRepository onMessage when session expires

### ISSUE THREAD

**Issue by kamal2311** [spring-projects/spring-session#578]
Using spring session 1.2.1.RELEASE with Redis 3.0.4.2. When attempting to catch the SessionDestroyedEvent, 

         @SuppressWarnings("unchecked")
	public void onMessage(Message message, byte[] pattern) {
		byte[] messageChannel = message.getChannel();
		byte[] messageBody = message.getBody();
		if (messageChannel == null || messageBody == null) {
			return;
		}

		String channel = new String(messageChannel);

		if (channel.startsWith(getSessionCreatedChannelPrefix())) {
			// TODO: is this thread safe?
			Map<Object, Object> loaded = (Map<Object, Object>) this.defaultSerializer
					.deserialize(message.getBody());
			handleCreated(loaded, channel);
			return;
		}

		String body = new String(messageBody);
		if (!body.startsWith(getExpiredKeyPrefix())) {
			return;
		}

The variable body has some junk characters before the actual value spring:session:.....
Is it stored in Redis incorrectly or retrieved incorrectly? Should an explicit charset be specified while storing and retrieving bytes?

**Comment by vpavic** [comment]
@kamal2311 Are you sure both scenarios use the same configuration, in particular the same serialization mechanism?

At first glance it looks like app on Windows machine uses `JdkSerializationRedisSerializer` whereas on Linux machine it uses `StringRedisSerializer`.

**Comment by kamal2311** [comment]
Yes I use the same source code on both the environments. I too think the problem is in serialization but my application code does not explicitly specify platform specific serialization. Problem seems to be some where in the framework code.

**Comment by rwinch** [comment]
I have to agree with @vpavic it appears that the Windows machine uses `JdkSerializationRedisSerializer` whereas on Linux machine it uses `StringRedisSerializer`.

Can you perhaps add a debug point to both `JdkSerializationRedisSerializer` and  `StringRedisSerializer` on the windows machine to see which is being used to save session ids?

**Comment by kamal2311** [comment]
The debug point showed that it was StringRedisSerializer that was used on the Windows machine.

**Comment by rwinch** [comment]
@kamal2311 Are you using the exact same Redis instance for both the Linux and Windows app?

**Comment by kamal2311** [comment]
Yes. It is the exact same redis instance and exact same spring boot app in
both environments.

**Comment by rwinch** [comment]
@kamal2311 Thanks for the response. When I said client I meant to ask if you are using the same client to view the keys. For example, are you using the same `redis-cli` on the same machine to view both the Linux inserted keys and the Windows inserted keys.

**Comment by kamal2311** [comment]
@rwinch oh sorry I missed the client part. Yes I am using the same redis-cli on the windows machine to view the Redis session keys.

**Comment by ysjiang4869** [comment]
@kamal2311 do you have any methods to solve this problem? I have the same question , and i can't get the session using redistemplate now

**Comment by tommyacm** [comment]
anybody has a solution for this issue?
